core/scripts/vrfv2/genvrfnum/main.go

```diff
@@ -0,0 +1,247 @@
+					fmt.Sprintf("%d", *cbGasLimit), fmt.Sprintf("%d", *numWords)) // cb gas limit, num words
+
+				preseedData := proof.PreSeedDataV2{
+					PreSeed:          preSeed,
+					BlockHash:        blockhash,
+					BlockNum:         *blockNum,
+					SubId:            *subID,
+					CallbackGasLimit: uint32(*cbGasLimit),
+					NumWords:         uint32(*numWords),
+					Sender:           sender,
+				}
+				finalSeed := proof.FinalSeedV2(preseedData)
+
+				record = append(record, finalSeed.String())
+
+				// generate proof
+				pf, err2 := key.GenerateProof(finalSeed)
+				helpers.PanicErr(err2)
+
+				record = append(record,
+					hex.EncodeToString(secp256k1.LongMarshal(pf.PublicKey)), // pub key
+					hex.EncodeToString(secp256k1.LongMarshal(pf.Gamma)),     // gamma
+					pf.C.String(), pf.S.String(), // c, s
+					pf.Seed.String(), pf.Output.String()) // seed, output
+
+				if len(record) != len(header) {
+					panic("record length doesn't match header length - update one of them?")
+				}
+				outChan <- record
+				numIters++
+			}
+			fmt.Println("genProofs worker wrote", numIters, "records to channel")
+		}
+
+		outFile, err := os.Create(*outfile)
+		defer outFile.Close()
+		helpers.PanicErr(err)
+
+		csvWriter := csv.NewWriter(outFile)
+		helpers.PanicErr(csvWriter.Write(header))
+		gather := func(outChan chan []string) {
+			for {
+				select {
+				case row := <-outChan:
+					helpers.PanicErr(csvWriter.Write(row))
+				case <-time.After(500 * time.Millisecond):
+					// if no work is produced in this much time, we're probably done
+					return
+				}
+			}
+		}
+
+		ranges := nonceRanges(1, uint64(*numCount), *numWorkers)
+
+		fmt.Println("nonce ranges:", ranges, "generating proofs...")
+
+		outC := make(chan []string)
+
+		for _, nonceRange := range ranges {
+			go genProofs(
+				nonceRange,
+				outC)
+		}
+
+		gather(outC)
+	case "verify-vrf-proofs":
+		cmd := flag.NewFlagSet("verify-vrf-proofs", flag.ExitOnError)
+		csvPath := cmd.String("p", "randomnumbers.csv", "path to csv file generated by gen-vrf-numbers")
+		numWorkers := cmd.Int("num-workers", runtime.NumCPU()-1, "number of workers to run verification")
+
+		helpers.ParseArgs(cmd, os.Args[2:])
+
+		numValid := &atomic.Int64{}
+		proofsChan := make(chan *vrfkey.Proof)
+
+		verify := func(inC chan *vrfkey.Proof) {
+			for {
+				select {
+				case pf := <-inC:
+					valid, err := pf.VerifyVRFProof()
+					helpers.PanicErr(err)
+					if !valid {
+						fmt.Println("proof", pf.String(), "is not valid", "total valid proofs:", numValid)
+						panic("found invalid proof")
+					}
+					numValid.Inc()
+				case <-time.After(250 * time.Millisecond):
+					return
+				}
+			}
+		}
+
+		for i := 0; i < *numWorkers; i++ {
+			go verify(proofsChan)
+		}
+
+		f, err := os.Open(*csvPath)
+		helpers.PanicErr(err)
+		defer f.Close()
+		reader := csv.NewReader(f)
+		_, err = reader.Read() // read the column titles
+		helpers.PanicErr(err)
+		for {
```

Comment:
So multiple go routines are set up to handle the incoming proofs, but then they're iterated through synchronously here?

Reply:
This loop is the producer - it writes the proofs to the channel and that channel is consumed by the workers running on separate goroutines.